sys/kern/src/task.rs

```diff
@@ -221,12 +221,44 @@ impl Task {
             // according to the task's region map... but fine with us.
             return true;
         }
+        // Forceably include DEVICE in the forbidden set, whether or not the
+        // caller thought about it.
         let forbidden = forbidden | RegionAttributes::DEVICE;
-        self.region_table().iter().any(|region| {
-            region.covers(slice)
-                && region.attributes.contains(desired)
-                && !region.attributes.intersects(forbidden)
-        })
+
+        // We need to be convinced that this slice is _entirely covered_ by
+        // regions with the right attributes. It _may_ span multiple regions,
+        // which is important since the build system can insert region
+        // boundaries in unpredictable locations in otherwise innocent programs.
+        // But the regions that are spanned must be perfectly contiguous.
+
+        let mut scan_addr = slice.base_addr;
+        let end_addr = slice.end_addr();
+
+        for region in self.region_table() {
+            if region.contains(scan_addr) {
+                // Make sure it's permissible!
+                if !region.attributes.contains(desired) || region.attributes.intersects(forbidden) {
+                    return false;
+                }
+
+                if end_addr < region.end_addr() {
+                    // We've exhausted the slice in this region, we don't have
+                    // to continue processing.
+                    return true;
+                }
+
+                // Continue scanning at the end of this region.
+                scan_addr = region.end_addr();
+            } else if region.end_addr() > scan_addr {
+                // We've passed our target address without finding regions that
+                // work!
+                break;
```

Comment:
very minor nit, not actually important: it's _slightly_ weird to me that we use `return false` _and_ `break`, both of which ultimately result in us returning `false` --- i would kinda prefer it if both of the cases where we return false used the same thing?

Reply:
Good point, fixing.